theano/tensor/basic.py

```diff
@@ -1434,7 +1407,7 @@ def c_code(self, node, name, inp, out, sub):
         x, axis = inp
         max, argmax = out
         fail = sub["fail"]
-        if NoneConst.equals(node.inputs[1]):
+        if NoneConst.equals(node.inputs[1]) or len(node.inputs[1].data) == node.inputs[0].ndim:
```

Comment:
The NoneConst case should not happen anymore.  However, you can't depend on the shape of the input to generate the C code since that can change without theano recompiling the C code.  If it is really necessary to have a numeric value for axis at compile time, then make is a parameter of the op, not an input.

If that becomes too much (after all the issue was not about fixing crazy things in the CPU code), you can also choose to leave the CPU code alone and perform the necessary conversions of the axis in the optimization that moves to the GPU.